File: ScoliosisCoach-iOS/ScoliosisCoach/Analysis/PostureRules.swift
import Foundation

struct PostureThresholds {
    let shoulderDiffPx: CGFloat
    let trunkAngleDeg: CGFloat

    static let `default` = PostureThresholds(shoulderDiffPx: 15, trunkAngleDeg: 3.0)
}

let ageThresholds: [String: PostureThresholds] = [
    "under12": PostureThresholds(shoulderDiffPx: 12, trunkAngleDeg: 2.5),
    "under15": PostureThresholds(shoulderDiffPx: 15, trunkAngleDeg: 3.0),
    "under18": PostureThresholds(shoulderDiffPx: 18, trunkAngleDeg: 3.5),
    "adult": PostureThresholds(shoulderDiffPx: 20, trunkAngleDeg: 4.0),
]

struct PostureResult {
    let shoulderAsymmetry: CGFloat
    let hipAsymmetry: CGFloat
    let trunkLeanAngle: CGFloat
    let headTilt: CGFloat
    let spineDeviation: CGFloat
    let armHangDiff: CGFloat
    let shoulderStatus: String
    let trunkStatus: String
}

func analyzePosture(
    landmarks: [[String: CGFloat]],
    imageSize: CGSize,
    ageGroup: String = "under15"
) -> PostureResult? {
    guard landmarks.count >= 29 else { return nil }

    let detector = PoseDetector.shared

    let lSh = detector.landmark(landmarks, at: 11) ?? .zero
    let rSh = detector.landmark(landmarks, at: 12) ?? .zero
    let lHip = detector.landmark(landmarks, at: 23) ?? .zero
    let rHip = detector.landmark(landmarks, at: 24) ?? .zero
    let nose = detector.landmark(landmarks, at: 0) ?? .zero
    let lWrist = detector.landmark(landmarks, at: 15) ?? .zero
    let rWrist = detector.landmark(landmarks, at: 16) ?? .zero

    let midShX = (lSh.x + rSh.x) / 2
    let midShY = (lSh.y + rSh.y) / 2
    let midHipX = (lHip.x + rHip.x) / 2
    let midHipY = (lHip.y + rHip.y) / 2

    let shoulderDiff = abs(lSh.y - rSh.y)
    let hipDiff = abs(lHip.y - rHip.y)

    // Trunk lean angle
    let dx = midShX - midHipX
    let dy = midHipY - midShY
    let trunkAngle = abs(atan2(dx, dy) * 180 / .pi)

    let headTilt = abs(nose.x - midShX)

    let spineDeviation = abs(midShX - midHipX)

    let leftArmLength = lSh.y - lWrist.y
    let rightArmLength = rSh.y - rWrist.y
    let armDiff = abs(leftArmLength - rightArmLength)

    // FIX #8: safe fallback instead of force-unwrap
    let thresholds = ageThresholds[ageGroup] ?? .default

    let shoulderStatus = shoulderDiff < thresholds.shoulderDiffPx ? "good" : "needs_improvement"
    let trunkStatus = trunkAngle < thresholds.trunkAngleDeg ? "good" : "needs_improvement"

    return PostureResult(
        shoulderAsymmetry: shoulderDiff.rounded(to: 2),
        hipAsymmetry: hipDiff.rounded(to: 2),
        trunkLeanAngle: trunkAngle.rounded(to: 2),
        headTilt: headTilt.rounded(to: 2),
        spineDeviation: spineDeviation.rounded(to: 2),
        armHangDiff: armDiff.rounded(to: 2),
        shoulderStatus: shoulderStatus,
        trunkStatus: trunkStatus
    )
}

extension CGFloat {
    func rounded(to places: Int) -> CGFloat {
        let divisor = pow(10.0, CGFloat(places))
        return (self * divisor).rounded() / divisor
    }
}

File: analysis/posture_rules.py
from .pose_detector import get_detector
import numpy as np

AGE_THRESHOLDS = {
    'under12': {'shoulder_diff': 12, 'trunk_angle': 2.5},
    'under15': {'shoulder_diff': 15, 'trunk_angle': 3.0},
    'under18': {'shoulder_diff': 18, 'trunk_angle': 3.5},
    'adult': {'shoulder_diff': 20, 'trunk_angle': 4.0}
}

def analyze_posture(landmarks, image_shape, age_group='under15'):
    if not landmarks:
        return {}
    
    # Key landmarks
    l_sh = get_detector().get_landmark_coords(landmarks, image_shape, 11)
    r_sh = get_detector().get_landmark_coords(landmarks, image_shape, 12)
    l_hip = get_detector().get_landmark_coords(landmarks, image_shape, 23)
    r_hip = get_detector().get_landmark_coords(landmarks, image_shape, 24)
    nose = get_detector().get_landmark_coords(landmarks, image_shape, 0)
    l_ankle = get_detector().get_landmark_coords(landmarks, image_shape, 27)
    r_ankle = get_detector().get_landmark_coords(landmarks, image_shape, 28)
    
    # Midpoints
    mid_sh = [(l_sh[0] + r_sh[0])/2, (l_sh[1] + r_sh[1])/2]
    mid_hip = [(l_hip[0] + r_hip[0])/2, (l_hip[1] + r_hip[1])/2]
    
    # Metrics
    shoulder_diff = abs(l_sh[1] - r_sh[1])
    hip_diff = abs(l_hip[1] - r_hip[1])
    
    # Trunk lean angle (coronal plane)
    trunk_angle = abs(np.arctan2(mid_sh[0] - mid_hip[0], mid_hip[1] - mid_sh[1]) * 180 / np.pi)
    
    # Head tilt
    head_tilt = abs(nose[0] - mid_sh[0])
    
    # Spine deviation
    spine_deviation = abs(mid_sh[0] - mid_hip[0])
    
    # Arm hang asymmetry
    l_wrist = get_detector().get_landmark_coords(landmarks, image_shape, 15)
    r_wrist = get_detector().get_landmark_coords(landmarks, image_shape, 16)
    arm_diff = abs((l_sh[1] - l_wrist[1]) - (r_sh[1] - r_wrist[1]))
    
    thresholds = AGE_THRESHOLDS.get(age_group, AGE_THRESHOLDS['under15'])
    
    # Status determination
    shoulder_status = "good" if shoulder_diff < thresholds['shoulder_diff'] else "needs_improvement"
    trunk_status = "good" if trunk_angle < thresholds['trunk_angle'] else "needs_improvement"
    
    return {
        'shoulder_asymmetry': round(shoulder_diff, 2),
        'hip_asymmetry': round(hip_diff, 2),
        'trunk_lean_angle': round(trunk_angle, 2),
        'head_tilt': round(head_tilt, 2),
        'spine_deviation': round(spine_deviation, 2),
        'arm_hang_diff': round(arm_diff, 2),
        'shoulder_status': shoulder_status,
        'trunk_status': trunk_status
    }
